Include/cpython/listobject.h

```diff
@@ -44,7 +44,11 @@ PyList_SET_ITEM(PyObject *op, Py_ssize_t index, PyObject *value) {
     PyListObject *list = _PyList_CAST(op);
     assert(0 <= index);
     assert(index < list->allocated);
+#ifdef Py_GIL_DISABLED
+    _Py_atomic_store_ptr_release(&list->ob_item[index], value);
```

Comment:
This doesn't feel right to me. Basically, `PyList_SET_ITEM` is only safe in certain circumstances during list initialization before the list is exposed to other threads. If the list is later exposed to other threads, it seems like there should be other synchronization involved to make the whole thing thread-safe.

Is there a reproducer for the Tsan reported race?

Reply:
I added a targetted test that repros it consistently under a TSAN build and we only need a relaxed load here.

Reply:
Ok, but I still think it makes more sense to modify `_PyList_AppendTakeRefListResize` to use `FT_ATOMIC_STORE_PTR_RELEASE` (or relaxed?) directly rather than modifying `PyList_SET_ITEM`.

`PyList_SET_ITEM` is used in ~2 places internally (`_PyList_AppendTakeRef` and `_PyList_AppendTakeRefListResize`) to modify existing lists while holding the lock. It's also used in many places to initialize newly created lists. The atomic operations make sense in the `_PyList_AppendTakeRef*` calls, but not really for the other uses of `PyList_SET_ITEM`.

I'm also not sure "relaxed" is strong enough for `_PyList_AppendTakeRef*` -- we generally use "release" when assigning to the entries of `ob_item`.

Reply:
One of the other places we're missing an atomic store is in `list_extend_fast` but there seems to be some actual real crashes when we're extending and iterating at the same time that I'm looking into.